### PR TITLE
Njavokhir/loading widget #27

### PR DESCRIPTION
**Pull request by NJavokhir**
- **Loading Percentage & Progress Bar:** Added real-time loading percentage and progress bar for image uploads during the raining effect.
- **Refactoring & Modularization:** Refactored code for readability, extracted image-fetching into a reusable module.
- **Latest Changes from `njavokhir/new-world-transition-#21` branch which is from PR #41 :** Merged updates from `njavokhir/new-world-transition-#21` for feature alignment and compatibility.

### CODE REVIEW

**Comment by NJavokhir** [comment]
@stravid87,  Merged updates from `njavokhir/new-world-transition-#21` which is PR number #41 . So, we can close #41. Also, refactoring and improving changes are here.

File: index.html [hunk 1]
```diff
@@ -7,7 +7,7 @@
     <title>Vite App</title>
   </head>
   <body>
-    <div id="app" class=""></div>
+    <div id="app"></div>
```
Comment (stravid87) [inline index.html:10]: Thank you for cleaning up.

File: package.json [hunk 2]
```diff
@@ -18,6 +18,7 @@
     "gsap": "^3.12.5",
     "layer8_interceptor": "^0.0.22",
     "locomotive-scroll": "^4.1.4",
+    "mitt": "^3.0.1",
```
Comment (stravid87) [inline package.json:21]: Tiny. Focused. Good choice.

File: src/assets/base.css [hunk 3]
```diff
@@ -6,3 +6,28 @@
...
+  display: block;
+  height: 100%;
+  width: 100%;
+  position: fixed;
+  top: 0; 
+  left: 0; 
+  animation-name: fadeinout;
+  animation-duration: 6s;
+  animation-iteration-count: 1;
+  pointer-events: none;
```
Comment (stravid87) [inline src/assets/base.css:20]: What is the purpose of this? "Pointer events none"

File: src/assets/base.css [hunk 4]
```diff
@@ -6,3 +6,28 @@
...
+
+canvas { 
+  display: block;
+  height: 100%;
+  width: 100%;
+  position: fixed;
+  top: 0; 
+  left: 0; 
+  animation-name: fadeinout;
+  animation-duration: 6s;
```
Comment (stravid87) [inline src/assets/base.css:18]: We will work on the transition next. 
I'm not sure what this does.... changing it doesn't seem to affect the frontend.

File: src/router/index.ts [hunk 5]
```diff
@@ -35,4 +36,17 @@ const router = createRouter({
   ]
 })
 
-export default router
+
+
+router.beforeEach((to, from, next) => {
+  if (to.name !== "home") {
+    triggerRainingEffect();
+    setTimeout(() => {
```
Comment (stravid87) [inline src/router/index.ts:44]: I like to think that setTimeouts are almost always cheating. We'll work on this next week.

File: src/utils/codeRainingEffect.ts [hunk 6]
```diff
@@ -1,104 +1,77 @@
...
-export const handleCodeRainingEffect = (value: boolean) => {
-  if (value) {
-    if (intervalId) return;   
-
-    loadingPercentage = 0;
-
-    // Create canvas
-    canvas = document.createElement('canvas');
-    document.body.appendChild(canvas);
+let percentage: number | 0 = 0;
```
Comment (stravid87) [inline src/utils/codeRainingEffect.ts:5]: why `number | 0`
is 0 not a `number`?

Comment (stravid87) [inline src/utils/codeRainingEffect.ts:5]:
```suggestion
let percentage: number = 0;
```

File: src/utils/useFetchImages.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        loadingPercentage.value = Math.round(((i + 1) / totalImages) * 100);
+        console.log(`Image ${i + 1} loaded in ${loadTime} ms`);
+        console.log("PERCENTAGE", loadingPercentage.value);
+
+        eventBus.emit('loading-percentage', loadingPercentage.value);
+      }
+
+      images.value = imgs;
+      isLoaded.value = true;
+      setTimeout(() => {stopRainingEffect()}, 1000)
```
Comment (stravid87) [inline src/utils/useFetchImages.ts:44]: We will deal with this next....

File: src/views/HiddenView.vue [hunk 8]
```diff
@@ -1,42 +1,19 @@
...
-      handleCodeRainingEffect(false);
-      images.value = imgs;
-      isLoaded.value = true;
-    }).then(()=>{
-      gsap
-    .timeline({
+const router = useRouter();
+const BACKEND_URL = import.meta.env.VITE_BACKEND_URL;
+
+// This gsap function should be utility, I will do this next...
```
Comment (stravid87) [inline src/views/HiddenView.vue:13]: Deal. 👍

File: src/views/HiddenView.vue [hunk 9]
```diff
@@ -45,80 +22,82 @@ const fetchImages = async () => {
...
       <section v-else>
-        <div>
-          <img :src="images[2].url" alt="image.name" />
-        </div>
-        <div>
-          <img :src="images[3].url" alt="image.name" />
-        </div>
-        <div>
-          <img :src="images[4].url" alt="image.name" />
+        <!-- `images.slice(2, 5)` is used for getting images from second to fifth, I will rewrite this when I use real images -->
```
Comment (stravid87) [inline src/views/HiddenView.vue:93]: 👍

File: src/views/HomeView.vue [hunk 10]
```diff
@@ -156,4 +84,13 @@ onMounted(() => {
   </main>
 </template>
 
-<style lang="scss"></style>
+<style lang="scss">
```
Comment (stravid87) [inline src/views/HomeView.vue:74]: lang scss?
```suggestion
<style>
```

**Review by stravid87 — APPROVED** [review]
Good work solving the loading widget challenge. Next up, the smooth fade in / fade out.

**Pull request merged**